### ISSUE TITLE
How can I properly store the best weights into a json file to use it later?

### ISSUE THREAD

**Issue by shinechou** [uber/horovod#54]
@alsrgv: If I use multiple GPUs, I tried to store the best weights to use it later but I couldn't get the results like what I could get using single GPU, what is the proper way to do it by using Horovod? Thanks a lot in advance.

**Comment by alsrgv** [comment]
@shinechou, you may need to adjust your learning rate when you use multiple GPUs - the simplest strategy is to just multiply your original learning rate by total # of GPUs.

If you're using Keras you can use https://keras.io/callbacks/#modelcheckpoint callback, and specify `save_best_only=True` in which case it will save best model in .h5 format, that you can subsequently load back with:

```python
model = keras.models.load_model('/path/to/model')
```

**Comment by shinechou** [comment]
@alsrgv: thank you very much. I did like what you said. Actually you helped me about how to adjust the learning rate. And I am using the "checkpoint = ModelCheckpoint(weightsfilepath, monitor='val_acc', verbose=1, save_best_only=True, mode='max')" to store the model with the weights. But I think maybe the problem comes from the loading model "loaded_model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])", here the optimizer is different from the one used during training ("tf.train.AdadeltaOptimizer"), how can I solve it? 

I have a question about the optimizer, if I use tensorflow optimizer I don't have Nadam optimizer (Keras does have), right?

Besides, when I plot the val_acc/val_loss when using multiple GPUs I got very strange values (higher than 1) like 

38656/38656 [==============================] - 25s - loss: 4.1219 - acc: 0.1017 - val_loss: 2.0161 - val_acc: 1.0590
Epoch 5/40
38656/38656 [==============================] - 25s - loss: 4.1219 - acc: 0.1025 - val_loss: 2.0081 - val_acc: 1.0671

How can I get the correct ones? I've already use the "        for metric in logs:
            avg_op = hvd.allreduce(K.constant(logs[metric]))
            logs[metric] = K.get_session().run(avg_op)"
to average the metrics from different GPUs. What went wrong here? Thanks again for all your help.

# Adjust learning rate based on number of GPUs (naive approach).
learning_rate = K.variable(1.0 * hvd.size())
print('\n\n learning_rate: ', learning_rate)
tf_opt = tf.train.AdadeltaOptimizer(learning_rate)  # Adadelta  0.01

# Add Horovod Distributed Optimizer.
tf_opt = hvd.DistributedOptimizer(tf_opt)

# Wrap with TF optimizer with Keras optimizer, pass learning rate.
opt = keras.optimizers.TFOptimizer(tf_opt)
opt.lr = learning_rate

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=opt,
              metrics=['accuracy'])

**Comment by alsrgv** [comment]
1. I'm not sure I understand your first problem - do you get an exception during loading the model?

2. TensorFlow seems to have Nadam optimizer - https://www.tensorflow.org/api_docs/python/tf/contrib/opt/NadamOptimizer

3. I'm very surprised to see accuracy > 1. While I don't know how that was possible, the learning rate you use with AdadeltaOptimizer seems to be very high. You start with # GPU * 1.0, while default LR for AdadeltaOptimizer is 0.001. What if you try LR = 0.001 * # GPUs?

Lastly, we're working on Horovod Keras API which would support native Keras optimizers, metric averaging, data partitioning and other goodies, along with examples how to use it - this should be coming sometime next week and hopefully will alleviate some of these problems, too.

**Comment by shinechou** [comment]
Really appreciate. Thank you.

(1) No, I didn't get exception. I can load the model but the inference result is very different from the one with single GPU. I think maybe also due to the strange val_acc values because I monitor it to store the model and weights.

(2) Yesterday I tried the optimizer as you mentioned but I got error at that time, now I try the learning rate of "0.001* #GPUs" and it works fine.

(3) I still got the similar strange val_acc values with LR = 0.001 * # GPUs.

38656/38656 [==============================] - 25s - loss: 1.8129 - acc: 0.4893 - val_loss: 0.9586 - **val_acc: 1.1829**
Epoch 22/40
38656/38656 [==============================] - 25s - loss: 1.8123 - acc: 0.4893 - val_loss: 1.9151 - val_acc: 0.2263

Really looking forward to your great work (Horovod Keras API and examples). Thanks again.

**Comment by alsrgv** [comment]
Very strange.  These versions seem very reasonable.

**Comment by shinechou** [comment]
@alsrgv: is it possibly caused by 2 different type of GPUs?

**Comment by alsrgv** [comment]
@shinechou, do you have 2 different types of GPUs on your servers, or between your home setup and at work?

Also, can you try this improved version of `MetricAverageCallback`? It names metrics specifically, so there's less chance that they will get mixed up by allreduce.
```python
class MetricAverageCallback(keras.callbacks.Callback):
    def __init__(self, device=''):
        self.variables = {}
        self.allreduce_ops = {}
        self.device = device

    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        for metric in logs:
            with tf.device(self.device):
                if metric not in self.variables:
                    with K.name_scope('MetricAverageCallback'):
                        self.variables[metric] = K.variable(logs[metric], name=metric)
                        self.allreduce_ops[metric] = hvd.allreduce(self.variables[metric])
                else:
                    K.set_value(self.variables[metric], logs[metric])
                logs[metric] = K.get_session().run(self.allreduce_ops[metric])
```

**Comment by shinechou** [comment]
@alsrgv: thank you very much. I have only single GPU at home so I couldn't try it. Tomorrow I'll borrow another GPU from work to try it and let you know. That's why I had two different types of GPUs when I tried it last time at home. Have a nice weekend.

**Comment by alsrgv** [comment]
@shinechou, you too. Let me know how it goes!

If the new version of `MetricAverageCallback` doesn't help, you can also try disabling NCCL by passing `device_dense="/cpu:0"' to `hvd.DistributedOptimizer` (see example below). If that helps, it means there may be a bug in NCCL related to dealing with different GPUs in the same ring.

```python
    # Add Horovod Distributed Optimizer, disable NCCL.
    opt = hvd.DistributedOptimizer(opt, device_dense='/cpu:0')
```

**Comment by shinechou** [comment]
@alsrgv: thanks again. I've tried the newer version of MetricAverageCallback with 2 GPUs at office (they worked well with your old version of MetricAverageCallback), I got errors as below,

----------------------------------------------------------------------------------------------------------------------------------
Epoch 1/50
Epoch 1/50
38624/38656 [============================>.] - ETA: 0s - loss: 5.2317 - acc: 0.0117WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_val_loss_0 [ready ranks: 1], MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_acc_0 [ready ranks: 0]
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_val_loss_0 [ready ranks: 1], MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_acc_0 [ready ranks: 0]
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_val_loss_0 [ready ranks: 1], MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_acc_0 [ready ranks: 0]
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_val_loss_0 [ready ranks: 1], MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_acc_0 [ready ranks: 0]
----------------------------------------------------------------------------------------------------------------------------------

and then I tried the " opt = hvd.DistributedOptimizer(opt, device_dense='/cpu:0')" trick as you pointed out, but still I got the similar error messaage,

-----------------------------------------------------------------------------------------------------
Epoch 1/50
Epoch 1/50
38624/38656 [============================>.] - ETA: 0s - loss: 5.2224 - acc: 0.012738624/38656 [============================>.] - ETA: 0s - loss: 5.2224 - acc: 0.0127WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_val_loss_0 [ready ranks: 1], MetricAverageCallback/HorovodAllreduce_MetricAverageCallback_loss_0 [ready ranks: 0]
------------------------------------------------------------------------------------------------------------------------

These two GPUs are the same, so I don't know why I got this strange error, could you please help me? Thanks in advance. I'll later also try 2 different GPUs to see what happens.

**Comment by alsrgv** [comment]
@shinechou, this confirms my suspicion that original issue was due to the different order of metrics in the dictionary. Here's a modified version of `MetricAverageCallback` that orders metrics, could you try it?

```python
class MetricAverageCallback(keras.callbacks.Callback):
    def __init__(self, device=''):
        self.variables = {}
        self.allreduce_ops = {}
        self.device = device

    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        for metric in sorted(logs.keys()):
            with tf.device(self.device):
                if metric not in self.variables:
                    with K.name_scope('MetricAverageCallback'):
                        self.variables[metric] = K.variable(logs[metric], name=metric)
                        self.allreduce_ops[metric] = hvd.allreduce(self.variables[metric])
                else:
                    K.set_value(self.variables[metric], logs[metric])
                logs[metric] = K.get_session().run(self.allreduce_ops[metric])
```

**Comment by shinechou** [comment]
@alsrgv: thank you so much. I've tred your latest version but I got the same error. But I think I figured it out, it is not caused by your code. If I use python 2.7 then everything is fine (even with your first version of MetricAverageCallback code), but things go wrong if I use python 3.5/6-x. Maybe tensorflow-gpu has a bug related to this, I don't know yet. Maybe you can give me some hints?

**Comment by alsrgv** [comment]
@shinechou, it's strange that new version doesn't work.  I was able to repro your issue on python3.4 with old code, but I can't repro the issue with `for metric in sorted(logs.keys())` modification.

**Comment by shinechou** [comment]
@alsrgv: thx a lot. I even didn't know that you repro. my problem using python 3.4. Indeed, it is weird, I tried several times, both at office and at home, python 2.7 always give me correct results but python 3.5 always give me wrong ones.

**Comment by alsrgv** [comment]
@shinechou, can you add this debug information to the callback:

```python
    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        # debug here
        print('Unsorted: ', logs.keys())
        print('Sorted: ', sorted(logs.keys()))
        # end of debug
        for metric in sorted(logs.keys()):
```

**Comment by shinechou** [comment]
@alsrgv: now it's going well with those two lines of code for debugging. I am really confused.

**Comment by alsrgv** [comment]
Closing this issue. Feel free to reopen if you have more questions.
